Stratis.Bitcoin/BlockPulling/LookaheadBlockPuller.cs

```diff
@@ -150,7 +150,11 @@ public int MaxBufferedSize
         /// <summary>Current number of bytes that unconsumed blocks are occupying.</summary>
         private long currentSize;
 
+        /// <summary>Lock object to protect access to <see cref="downloadedCounts"/>.</summary>
+        private object downloadedCountsLock = new object();
+
         /// <summary>Maintains the statistics of number of downloaded blocks. This is used for calculating new actualLookahead value.</summary>
+        /// <remarks>All access to this object has to be protected by <see cref="downloadedCountsLock"/>.</remarks>
```

Comment:
You could ToList() on the dictionary and work with the copy of the list in that point in time.
But that's fine its actually way more clear what is happening when using a lock() then a concurrent collection, though I would try to avoid calling i to methods from inside a lock (ah well its a static method in this case so fine).